src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Unix/HelpersOpenSsl.cs

```diff
@@ -0,0 +1,203 @@
+            {
+                Interop.Crypto.CheckValidOpenSslHandle(oidAsn1);
+                return Interop.Crypto.GetOidValue(oidAsn1);
+            }
+        }
+
+        public static ContentInfo GetEmbeddedContent(this SafeCmsHandle cmsHandle)
+        {
+            string oid;
+            Oid contentType;
+            byte[] content;
+
+            using (SafeSharedAsn1ObjectHandle oidAsn1 = Interop.Crypto.CmsGetEmbeddedContentType(cmsHandle))
+            {
+                Interop.Crypto.CheckValidOpenSslHandle(oidAsn1);
+                oid = Interop.Crypto.GetOidValue(oidAsn1);
+            }
+
+            contentType = new Oid(oid);
+
+            using (SafeSharedAsn1OctetStringHandle encodedContent = Interop.Crypto.CmsGetEmbeddedContent(cmsHandle))
+            {
+                // encodedContent can be a null pointer if there is no content. In this case the content should be set
+                // to an empty byte array. 
+                if (encodedContent.IsInvalid)
+                    content = new byte[0];
+                else
+                    content = Interop.Crypto.GetAsn1StringBytes(encodedContent);
+            }
+
+            return new ContentInfo(contentType, content);
+        }
+
+        public static X509Certificate2Collection GetOriginatorCerts(this SafeCmsHandle cmsHandle)
+        {
+            X509Certificate2Collection origCertCollection = new X509Certificate2Collection();
+
+            using (SafeX509StackHandle origCertsPtr = Interop.Crypto.CmsGetOriginatorCerts(cmsHandle))
+            {
+                // origCertsPtr might be a nullptr as originator certs are optional, but in this case,
+                // GetX509StackFieldCount will just return 0 and behavior will be as expected. 
+                int certCount = Interop.Crypto.GetX509StackFieldCount(origCertsPtr);
+                for (int i = 0; i < certCount; i++)
+                {
+                    IntPtr certRef = Interop.Crypto.GetX509StackField(origCertsPtr, i);
+                    Interop.Crypto.CheckValidOpenSslHandle(certRef);
+                    X509Certificate2 copyOfCert = new X509Certificate2(certRef);
+                    origCertCollection.Add(copyOfCert);
+                }
+            }
+
+            return origCertCollection;
+        }
+
+        public static AlgorithmIdentifier ReadEncryptionAlgorithm(this DerSequenceReader encodedCms)
+        {
+            // According to RFC2630, the enveloped type we have support for has the following structure
+            // EnvelopedData ::= SEQUENCE {
+            //     version CMSVersion,
+            //     originatorInfo[0] IMPLICIT OriginatorInfo OPTIONAL,
+            //     recipientInfos RecipientInfos,
+            //     encryptedContentInfo EncryptedContentInfo,
+            //     unprotectedAttrs[1] IMPLICIT UnprotectedAttributes OPTIONAL }
+            // In the GetVersion method we already consumed the sequence and the verison
```

Comment:
Since you've removed the GetVersion method this comment is a bit out of date.  But doing the skip in ReadEncryptionAlgorithm feels like a bit too much magic to me.  I'd either move this to the caller, or move it to a dedicated method (SkipOriginatorInfo) called by the caller; but I wouldn't expect ReadEncryptionAlgorithm to need to change should we ever have to parse originatorInfo ourselves.